src/app/index.js

```diff
@@ -0,0 +1,234 @@
+import fsApi from 'fs';
+import path from 'path';
+import events from 'events';
+
+import config from 'config';
+
+import * as history from './history/index.js';
+import fetch from './fetcher/index.js';
+import filter from './filter/index.js';
+import loadServiceDeclarations from './loader/index.js';
+
+const fs = fsApi.promises;
+const __dirname = path.dirname(new URL(import.meta.url).pathname);
+const SERVICE_DECLARATIONS_PATH = path.resolve(__dirname, '../../', config.get('serviceDeclarationsPath'));
+
+export const AVAILABLE_EVENTS = [
+  'startTrackingChanges',
+  'endTrackingChanges',
+  'startRefiltering',
+  'endRefiltering',
+  'snapshotRecorded',
+  'firstSnapshotRecorded',
+  'noSnapshotChanges',
+  'versionRecorded',
+  'firstVersionRecorded',
+  'noVersionChanges',
+  'changesPublished',
+  'applicationError',
+  'documentUpdateError',
+  'documentFetchError',
+  'recordRefilterError',
+];
+
+export default class CGUs extends events.EventEmitter {
+  get serviceDeclarations() {
+    return this._serviceDeclarations;
+  }
+
+  async init() {
+    if (!this.initialized) {
+      this._serviceDeclarations = await loadServiceDeclarations(SERVICE_DECLARATIONS_PATH);
+      this.initialized = Promise.resolve();
+    }
+
+    return this.initialized;
+  }
+
+  attach(listener) {
+    AVAILABLE_EVENTS.forEach(event => {
+      const handlerName = `on${event[0].toUpperCase()}${event.substr(1)}`;
+
+      if (Object.prototype.hasOwnProperty.call(listener, handlerName)
+      || Object.prototype.hasOwnProperty.call(listener.constructor.prototype, handlerName)) {
```

Comment:
I don't understand why we manually check `hasOwnProperty` on the parameter and its constructor instead of relying on the prototype chain navigation 🤔

Reply:
[live discussion] We'll use `[]` property existence check instead.